app/ducks/walletActions.js

```diff
@@ -368,12 +368,31 @@ async function parseInputsOutputs(net, tx) {
       covValue += output.value;
     }
 
+    if (covenant.action == 'FINALIZE') {
+
+      // isSender if there exists an input with value 0 and own address
+      const isSender = tx.inputs.findIndex(x => x.value === 0 && x.path !== null) !== -1;
+
+      // containsOwnPayment if there exists an input with non-zero value and is own address
+      const containsOwnPayment = tx.inputs.findIndex(x => x.value > 0 && x.path !== null) !== -1;
+
+      // If sender and has not paid, then it is Finalized with Payment by receiver: show +ve amount
+      // If receiver and has paid, then it is Finalized with Payment: show -ve amount
+      // Else, it is a simple Finalize: show 0 for both sender and receiver
+      if (isSender && !containsOwnPayment) {
+        covValue = tx.outputs.reduce((sum, op) => op.path ? (sum+op.value) : sum, 0);
+      } else if (!isSender && containsOwnPayment) {
+        covValue = -tx.outputs.reduce((sum, op) => !op.path ? (sum+op.value) : sum, 0);
+      } else {
+        covValue = 0;
+      }
+    }
+
     // Renewals and Updates have a value, but it doesn't
     // affect the spendable balance of the wallet.
     if (covenant.action === 'RENEW' ||
       covenant.action === 'UPDATE' ||
-      covenant.action === 'TRANSFER' ||
-      covenant.action === 'FINALIZE') {
+      covenant.action === 'TRANSFER') {
```

Comment:
I could be wrong but I think if the wallet is receiving payment for a FINALIZE, then the payment value might just be equal to `totalValue` calculated above. I'm not sure if that'll work for the buyer though (who is sending money) but might save you some of the logic in the blob L371-390

Reply:
Yup, this works for the sender and I initially tried this (but outside the for loop because totalValue won't be complete inside the loop).

```
if (covAction) {
    return {
      ...covData,
      fee: tx.fee,
      value: covenant.action === 'FINALIZE' ? totalValue : covValue, // <- instead of just covValue
    };
  }
```

Like you said, this only works for the seller, not buyer. There's also a problem of identifying if the value is +ve or -ve.